fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/SimpleBufferBuilder.java

```
@@ -0,0 +1,456 @@
+package net.minecraftforge.fml.earlydisplay;
+
+import org.lwjgl.system.MemoryStack;
+import org.lwjgl.system.MemoryUtil;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+import static org.lwjgl.opengl.GL32C.*;
+
+/**
+ * A very simple, Mojang inspired BufferBuilder.
+ * <em>This has been customized for 2d rendering such as text and simple planar textures</em>
+ * <p>
+ * Not bound to any specific format, ideally should be held onto for re-use.
+ * <p>
+ * Can be used for 'immediate mode' style rendering using {@link #draw()}, or
+ * upload to external vertex arrays for proper instancing using {@link #finishAndUpload()}.
+ * <p>
+ * This is a Triangles only buffer, all data uploaded is in Triangles.
+ * Quads are converted to triangles using {@code 0, 1, 2, 0, 2, 3}.
+ * <p>
+ * Any given {@link Format} should have its individual {@link Element} components
+ * buffered in the order specified by the {@link Format},
+ * followed by an {@link #endVertex()} call to prepare for the next vertex.
+ * <p>
+ * It is illegal to buffer primitives in any format other than the one specified to
+ * {@link #begin(Format, Mode)}.
+ *
+ * @author covers1624
+ */
+public class SimpleBufferBuilder implements Closeable {
+
+    private static final MemoryUtil.MemoryAllocator ALLOCATOR = MemoryUtil.getAllocator(false);
+
+    private static final int[] VERTEX_ARRAYS = new int[Format.values().length];
+    private static final int[] VERTEX_BUFFERS = new int[Format.values().length];
+    private static final int[] ELEMENT_BUFFERS = new int[Format.values().length];
+
+    static {
+        Arrays.fill(VERTEX_ARRAYS, -1);
```

glDelete* silently ignores zeros, should be used instead of `-1` to flag as unallocated